#Ajuste de redes nueronales
Diego Rosas A01634154

In [5]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPRegressor, MLPClassifier
from sklearn.model_selection import KFold, GridSearchCV, cross_val_predict
from sklearn.metrics import mean_squared_error, mean_absolute_error, classification_report

In [6]:
df = pd.read_csv('crime_data.csv')

In [7]:
#Debido a que mi matrícula (A01634154) termina en 4, trabajaré con Variable dependiente MR, variables independientes M, W, H y P
columnas_a_eliminar = ['VR', 'S']
df = df.drop(columnas_a_eliminar, axis=1)
df.head()

State    MR     M     W     H     P
0    AK   9.0  41.8  75.2  86.6   9.1
1    AL  11.6  67.4  73.5  66.9  17.4
2    AR  10.2  44.7  82.9  66.3  20.0
3    AZ   8.6  84.7  88.6  78.7  15.4
4    CA  13.1  96.7  79.3  76.2  18.2

In [8]:
x = df[['M', 'W', 'H', 'P']].values
y = df['MR'].values

#1. Evalúa con validación cruzada un modelo pereceptrón multicapa para las variables que se te asignaron para este ejercicio.

In [9]:
features = x
n_features = len(x)

# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 5
kf = KFold(n_splits=n_folds, shuffle = True)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):

    # Training phase
    x_train = x[train_index, :]
    y_train = y[train_index]

    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x[test_index, :]
    y_test = y[test_index]

    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE

    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)


print('MSE:', np.average(mse_cv), '  MAE:', np.average(mae_cv))

MSE:  157.30194190317334
MAE:  8.32028389893163
mse =  386.96939856425166
mae =  8.729980266135009
mse =  5.632651049051328
mae =  1.9866257103284481
mse =  46.75381716128403
mae =  5.155372291071769
mse =  5.922715548836769
mae =  2.050038777274653
mse =  5.757572837795588
mae =  2.031726533244817
MSE: 90.20723103224387   MAE: 3.99074871561094


#2. Agrega al conjunto de datos columnas que representen los cuadrados de las variables predictoras (por ejemplo, M2, W2), así como los productos entre pares de variables (por ejemplo, PxS, MxW). Evalúa un modelo perceptrón multicapa para este nuevo conjunto de datos.

In [10]:
df['M2'] = df['M'] ** 2
df['W2'] = df['W'] ** 2
df['H2'] = df['H'] ** 2
df['P2'] = df['P'] ** 2

df['MxW'] = df['M'] * df['W']
df['MxH'] = df['M'] * df['H']
df['MxP'] = df['M'] * df['P']

df['WxH'] = df['W'] * df['H']
df['WxP'] = df['W'] * df['P']

df['HxP'] = df['H'] * df['P']

x = df[['M', 'W', 'H', 'P','M2', 'W2', 'H2', 'P2', 'MxW', 'MxH', 'MxP', 'WxH', 'WxP', 'HxP']].values
y = df['MR'].values

In [11]:
features = x
n_features = len(x)

# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 5
kf = KFold(n_splits=n_folds, shuffle = True)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):

    # Training phase
    x_train = x[train_index, :]
    y_train = y[train_index]

    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x[test_index, :]
    y_test = y[test_index]

    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE

    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)


print('MSE:', np.average(mse_cv), '  MAE:', np.average(mae_cv))

MSE:  13.492052123791145
MAE:  2.582108312539482
mse =  109306.88412545664
mae =  320.25046789326376
mse =  150683.88556175467
mae =  340.08492441126435
mse =  6211.5793353787085
mae =  64.57098603465911
mse =  6640.738684525662
mae =  37.10952318850011
mse =  5442.5577790933985
mae =  30.847376451242496
MSE: 55657.129097241814   MAE: 158.57265559578596


#3. Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:

**¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?**

El modelo perceptrón multicapa no es nada efectivo, ya que como podemos observar en ambas ejecuciones (con variables normales y agregando sus cuadrados y multiplicaciones de pares), los errores MSE y MAE son sumamente altos lo cual nos indica que es sumamente probable que la clasificación en este método tenga errores.

**¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué?**
Analizando ambos modelos, ambos presentan errores y ninguno es sumamente bueno o perfecto, sin embargo, si se tuviera que elegir alguno sería el lineal sin duda alguna, ya que presenta mucho menos errores que el de perceptrón.

#Problema 2

In [12]:
# Load data set. Trabajé con el dataset 'M_3.txt' debido a que el penúltimo número de mi matrícula A01634154 es 5
data = np.loadtxt("M_3.txt")
x = data[:,2:]
y = data[:,0]

#1. Evalúa un modelo perceptrón multicapa con validación cruzada utilizando al menos 5 capas de 20 neuronas.

In [13]:
features = x
n_features = len(x)

################################################################################
# Train model with all observations (five layers of 20 neurons)
################################################################################
clf = MLPClassifier(hidden_layer_sizes=(20, 20, 20, 20, 20), max_iter=10000)
clf.fit(x, y)

################################################################################
# Evaluate model (five layers of 20 neurons)
################################################################################
y_pred = cross_val_predict(MLPClassifier(hidden_layer_sizes=(20, 20, 20, 20, 20), max_iter=10000), x, y)
print(classification_report(y, y_pred))


              precision    recall  f1-score   support

         1.0       0.97      0.97      0.97        90
         2.0       0.92      0.96      0.94        90
         3.0       0.84      0.91      0.87        90
         4.0       0.99      0.99      0.99        90
         5.0       0.99      0.94      0.97        90
         6.0       0.87      0.87      0.87        90
         7.0       0.93      0.86      0.89        90

    accuracy                           0.93       630
   macro avg       0.93      0.93      0.93       630
weighted avg       0.93      0.93      0.93       630



#2. Evalúa un modelo perceptrón multicapa con validación cruzada, pero encontrando el número óptimo de capas y neuronas de la red.

In [14]:
################################################################################
# Optimal number of layers and neurons
################################################################################
num_layers = np.arange(1, 20, 5)
num_neurons = np.arange(10, 110, 20)

layers = []

for l in num_layers:
    for n in num_neurons:
        layers.append(l*[n])

clf = GridSearchCV(MLPClassifier(max_iter=10000), {'hidden_layer_sizes': layers}, cv = 5)
clf.fit(x, y)
print(clf.best_estimator_)

MLPClassifier(hidden_layer_sizes=[90], max_iter=10000)


#3. Prepara el modelo perceptrón multicapa:

Opten los hiperparámetros óptimos de capas y neuronas de la red.

Con los hiperparámetros óptimos, ajusta el modelo con todos los datos.

In [15]:
################################################################################
# Evaluate model finding the optimal number of layers and neurons
################################################################################

clf = GridSearchCV(MLPClassifier(max_iter=10000), {'hidden_layer_sizes': layers}, cv = 5)
y_pred = cross_val_predict(clf, x, y, cv = 5)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

         1.0       0.95      0.99      0.97        90
         2.0       0.98      0.92      0.95        90
         3.0       0.95      0.88      0.91        90
         4.0       0.97      1.00      0.98        90
         5.0       1.00      0.94      0.97        90
         6.0       0.87      0.92      0.90        90
         7.0       0.94      0.99      0.96        90

    accuracy                           0.95       630
   macro avg       0.95      0.95      0.95       630
weighted avg       0.95      0.95      0.95       630



#4. Contesta lo siguiente

**¿Observas alguna mejora importante al optimizar el tamaño de la red? ¿Es el resultado que esperabas? Argumenta tu respuesta.**

Al buscar el número ideal de capas y neuronas, se observa cierta mejora, sin embargo, la diferencia no es mucha. Analizando los resultados, se pasó de una precisión de 0.93 a 0.95, lo cual si representa una mejora (que es lo que se esperaba al optimizar) pero la diferencia entre ambos no es mucha, sin embargo, al representar una mejora puede ser de utilidad.

**¿Qué inconvenientes hay al encontrar el tamaño óptimo de la red? ¿Por qué?**

El único inconveniente que puede existir son los tiempos de ejecución, ya que personalmente al ejecutar el código para encontrar el número ideal de capas y neuronas y al evaluar y ajustar este modelo ideal, el tiempo de ejecución fue bastante prolongado, pero se logró obtener un mejor resultado.